### PR TITLE
Module exp for craft

### PR DESCRIPTION
**Pull request by vuong177**
## Module exp for craft

What in exp module : 
 - Save whitelist in kvstore, it stores informations about DAO members and theirs maximum amount.
 - Param : 
   - DaoAccount : is DAO address, MsgMintAndAllocateExp only execute when signer is DaoAccount
  
   - MaxMintExp: is maxium amount can be allocate for one member in DAO
   - Denom :  denom will be mint (default is exp)
   - ExpPrice : price EXP base in DAO wallet.
 - Msgs : 
   - MsgMintAndAllocateExp : only signed by DAO, mint exp and transfer it to a member address
   
   - MsgBurnAndRemoveMember : can be signed by anyone in DAO, burn their exp and remove them from whitelist
   - MsgJoinDao: only signed by gov module account, when executed it adds information about members and their maximum exp to whitelist.  User can select the way they buy exp in proposal :
     - Transfer non-ibc asset (like ETH, BTC) to DAO wallet to join DAO (need DAO confirm and send TX mint) 
 
     - Transfer ibc-asset to  exp module, module confirm user tx and mint exp for them (not need DAO confirm).
   - MsgFundExp : DAO need fund ibc token to exp module for burn proccess.
   - MsgSpendIBCTokenForExp : this msg for user (who create proposal join DAO and passed), they send their ibc token to module address. And exp module will mint exp for them.

### CODE REVIEW

**Comment by nguyenvuong1122000** [comment]
What in exp module : 
 - Save whitelist in kvstore, it it stores information abount DAO member and maximum they can mint
 - Param : 
   - DaoAccount : is DAO address, msg only execute when signer is DaoAccount
   - MaxMintExp: is maxium amount can be allocate for one member in Dao
   - Denom :  denom will be mint (default is exp)
 - Msgs : 
   - MsgMintAndAllocateExp : only execute by DAO, mint exp and transfer it to a member address
   - MsgBurnAndRemoveMember : can be execute by anyone in DAO, burn their exp and remove they from whitelist
   - MsgJoinDao: only execute by gov module, when execute it add infomation about member and maximum exp they can mint to whitelist.

**Comment by nguyenvuong1122000** [comment]
- Working flow when mint exp :
  - User makes proposal to join DAO
  - DAO votes yes / no
  - if yes, User now whitelisted into DAO

  - User sends eth to DAO eth wallet
  - User tells DAO he sent it (can't make automate because asset not on chain)
  - DAO now makes Tx to give user XXXXX exp
  - DAO members have to sign it because it is multisig
  - now user has EXP
 
- Working flow when burn exp :
  - Member makes tx "BurnExpAndExitDao"
 
  - When tx in chain, chain check VestingPeriod (lock exp time).
  - If passed, burn member coin and remove member from whitelist

**Comment by nguyenvuong1122000** [comment]
@Reecepbcups, i think no problem with UST (or any ibc-asset). It can be a another choice for user  : 
```
User can select they way we buy exp in proposal :
 - Transfer non-ibc asset (like ETH, BTC) to DAO wallet to join DAO (need DAO confirm and send TX mint) 
 - Transfer ibc-asset to a exp module, module confirm user tx and mint exp for them (not need DAO confirm).
```

**Comment by nguyenvuong1122000** [comment]
Yes, if use ibc-asset, we can do this. Burn exp will only receive UST. I think it's OK

**Comment by nguyenvuong1122000** [comment]
@Reecepbcups, i think exp price can be query on chain by using band protocol (It can query ETH, BTC, and other token price). But what we will do with NFT asset. How we calculate valuation of DAO NFT ?

**Comment by Reecepbcups** [comment]
> @Reecepbcups, what will we do with NFT asset. How we calculate valuation of DAO NFT ? It like auction property and don't have price until selled.

If the DAO purchased the NFT, we use that sell price. Else it would be the ETH floor price if it was given to us AFAIK.
Price query would need to be off chain for that, which is how Cabcon desires

**Comment by nguyenvuong1122000** [comment]
@Reecepbcups pls take a look

**Comment by nguyenvuong1122000** [comment]
I'm changing kv store logic. Please don't merge.

**Comment by faddat** [comment]
Vuong, at the top of this PR can you please give the best possible narrative as to its function?

Do you need any feedback on it from anyone?

Thank you

**Review by faddat — APPROVED** [review]
Vuong,

so I read through this from a functional prespective and it really seems to cover all of the bases.  On may 1-7 tom and I will be doing a product review / front end / pre-launch planning thing.  You think this will be good to go?

**Comment by vuongnguyen17** [comment]
> Vuong,
> 
> so I read through this from a functional prespective and it really seems to cover all of the bases. On may 1-7 tom and I will be doing a product review / front end / pre-launch planning thing. You think this will be good to go?

i think it's good.  But this module need a litle change in kv store logic, i'll try make it complete in this month.

**Comment by vuongnguyen17** [comment]
Do you need details document for module ?

**Comment by vuongnguyen17** [comment]
@Reecepbcups do you pull newest code, i tested and did't got that errors.

**Comment by Reecepbcups** [comment]
> @Reecepbcups do you pull newest code ? i tested and did't get that errors.

commit 56966e1625e689a8ace63168d5f8c43a42f40824
Date:   Thu Apr 28 03:37:45 2022 +0700
"remove println"

File: test_proposal.json [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+  "messages": [
+    {
+      "@type": "/craft.exp.v1beta1.MsgJoinDaoByIbcAsset",
+      "join_address": "craft1hj5fveer5cjtn4wd6wstzugjfdxzl0xp86p9fl",
+      "gov_address": "craft10d07y265gmmuvt4z0w9aw880jnsr700jm5qjn0",
+      "amount": "1000000"
+    }
+  ],
+  "metadata": "4pIMOgIGx1vZGU=", 
+  "deposit": "100000000000stake"
```
Comment (Reecepbcups) [inline test_proposal.json:11]: Can you explain this? - 
If you own EXP you are apart of the DAO already. EXP (stake) is not transferable.

File: x/exp/keeper/exp_request.go [hunk 2]
```diff
@@ -0,0 +1,371 @@
...
+	return
+}
+
+// not good logic need modify
+func (k ExpKeeper) GetBurnRequest(ctx sdk.Context, accAddress sdk.AccAddress) (types.BurnRequest, error) {
+	store := ctx.KVStore(k.storeKey)
+
+	if store.Has(types.GetBurnRequestAddressBytes(int(types.StatusCompleteRequest), accAddress)) {
+		return k.GetBurnRequestByKey(ctx, types.GetBurnRequestAddressBytes(int(types.StatusCompleteRequest), accAddress))
+	}
```
Comment (Reecepbcups) [inline x/exp/keeper/exp_request.go:131]: a lot of duplicate code and double checking, should also use a variable like:
```go
ab := types.GetBurnRequestAddressBytes(int(types.StatusCompleteRequest), accAddress)
```

GetBurnRquestByKey ( line 130) also already as a store.Has() check, so line 129 is a duplicate of what 130 already does

Is minor but can def be improved to reduce undeeded store queries

File: test_proposal.json (hunk 1 above)
Comment (vuong177) [reply test_proposal.json:11]: Oh, this only for test bash. I mean 'stake' here is transable denom like 'ucraft'.

File: x/exp/keeper/exp_request.go (hunk 2 above)
Comment (vuong177) [reply x/exp/keeper/exp_request.go:131]: Oh yep, it's my bad logic.

Comment (vuong177) [reply x/exp/keeper/exp_request.go:132]: @Reecepbcups, i'll change logic in kvstore again for better performance

**Comment by faddat** [comment]
Hi, I am thinking about merging this in advance of a new testnet, what do you think?

**Comment by vuong177** [comment]
> Hi, I am thinking about merging this in advance of a new testnet, what do you think?

It's good. We implement most of
 logic in document `boring.pdf`. Ready to next testnet.

**Comment by faddat** [comment]
That's perfect, Tom and I want to start testing the module's functionality and thinking about ui this week. 

Thank you!

**Comment by faddat** [comment]
@Reecepbcups you did a good job reviewing.

**Comment by faddat** [comment]
Might have some decimal math errors here, if you can clean those up, I think we are ready for what I professional described as the `janky, feature complete testnet`

**Comment by vuong177** [comment]
> Might have some decimal math errors here, if you can clean those up, I think we are ready for what I professional described as the `janky, feature complete testnet`

i fixed. This bug because new update upsteam sdk.

**Comment by vuong177** [comment]
@faddat do you know why ci/cd build failded ?  I can build binary in my local computer, but ci/cd still failed.

**Pull request merged**